### DISCUSSION TITLE
punktf v4 - Go small go home

### DISCUSSION THREAD

**Discussion by Shemnei** [Shemnei/punktf#128] [Announcements]
Version v4 will be big rework of some core punktf features with the big goal on relying more on external crates to reduce the code base and make it more maintainable.

### Remove

- Drop support for `json` profiles, as this will reduce the test surface. Compared to `yaml` they are way to busy/cluttered anyways

### Additions

- Add integration tests, example tests and unit tests; Basically just __more tests__
- Integration for en/decryption? (still not sure if punktf should handle that; #24). If we add this feature, we probably also need to add build-in git support to be able to encrypt files before commiting to git and decrypting after a checkout.
- __FINISH DOCUMENTATION__ (#126)
- __ADD EXAMPLE REPOSITORIES__
- Version the profile definitions (e.g. `version: "2"`) so we are able to distinguish between them in the future to allow migrations/support (#8)
- Support for linux file permission, mostly meant to make files executable (#15)

### Rework

- Generic template engine interface
- Generic dotfile merge interface
- Replace custom templating engine with external crate (#121, #122, #53). One of those:
  - [tera](https://lib.rs/crates/tera)
  - [handlebars](https://lib.rs/crates/handlebars)
  - [minijinja](https://lib.rs/crates/minijinja)
- Custom variable syntax (env, profile, dotfile) will be merged into just environment variables
- `template` option will be replaced by files having the extension of the template engine (e.g. `test.txt` => `test.txt.tera`; Might need a way to "escape" this). This will have a few benefits:
  - Don't have to treat every file in a dotfile directory a template if the flag is set (performance)
  - The engine to use is directly related to the file, which if we want, allows us to use multiple engines in parallel
  - Syntax highlighting for editors
- Improve test/dry-runs by keeping track what files would be deployed to allow checking for overwrites, targets/sources of symlinks would exist

This will be a breaking change and no support for older versions will be provided, meaning repositories will have to be manually ported to the new version.
There will probably be a check if the profile/dotfile is using old/deprecated parts which will trigger a warning and stop the deployment.

I am open to other suggestion.

**Comment by michidk** [comment]
Please don't drop yaml. Nobody wants to configure anything in json directly

**Reply by Shemnei** [reply]
It's the other way around, i want to drop support for json and only keep yaml (it's under the remove section). After re-reading it, you are right, it might need some clearer wording 😓

**Reply by michidk** [reply]
ah lol

**Comment by michidk** [comment]
The scope might be too big, would recommend to work in increments
